pkg/controller/deployment/deployment_controller.go

```diff
@@ -306,6 +317,50 @@ func (dc *DeploymentController) handleErr(err error, key interface{}) {
 	dc.queue.Forget(key)
 }
 
+// classifyReplicaSets uses NewReplicaSetControllerRefManager to classify ReplicaSets
+// and adopts them if their labels match the Deployment but are missing the reference.
+// It also removes the controllerRef for ReplicaSets, whose labels no longer matches
+// the deployment.
+func (dc *DeploymentController) classifyReplicaSets(deployment *extensions.Deployment) error {
+	rsList, err := dc.rsLister.ReplicaSets(deployment.Namespace).List(labels.Everything())
+	if err != nil {
+		glog.Errorf("Error getting ReplicaSets for deployment %q: %v", deployment.Name, err)
+		return err
+	}
+
+	deploymentSelector, err := unversioned.LabelSelectorAsSelector(deployment.Spec.Selector)
+	if err != nil {
+		return fmt.Errorf("deployment %s/%s has invalid label selector: %v", deployment.Namespace, deployment.Name, err)
+	}
+	cm := controller.NewReplicaSetControllerRefManager(dc.rsControl, deployment.ObjectMeta, deploymentSelector, getDeploymentKind())
+	matchesAndControlled, matchesNeedsController, controlledDoesNotMatch := cm.Classify(rsList)
+	for _, replicaSet := range matchesNeedsController {
+		err := cm.AdoptReplicaSet(replicaSet)
+		// continue to next RS if adoption fails.
+		if err != nil {
+			// If the RS no longer exists, don't even log the error.
+			if !errors.IsNotFound(err) {
+				utilruntime.HandleError(err)
+			}
+		} else {
+			matchesAndControlled = append(matchesAndControlled, replicaSet)
+		}
+	}
+	// remove the controllerRef for the RS that no longer have matching labels
+	var errlist []error
+	for _, replicaSet := range controlledDoesNotMatch {
```

Comment:
I believe we should discourage selectors changes in controllers but allow changes to enable advanced use-cases (Blue-Green deployments is one). That being said I am not sure what changing a label selector for an app means semantically for the app. Should we treat a change to a new non-overlapping selector (non-overlapping to its previous self) , a new history for the Deployment (what happens now probably by accident) or should we continue maintaining the old history which doesn't match with the new selector anymore? 

cc: @smarterclayton @bgrant0607